Source/EasyNetQ/Producer/PublishConfirmationListener.cs

```diff
@@ -0,0 +1,103 @@
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EasyNetQ.Events;
+using EasyNetQ.Internals;
+using RabbitMQ.Client;
+
+namespace EasyNetQ.Producer
+{
+    public class PublishConfirmationListener : IPublishConfirmationListener
+    {
+        private readonly CancellationTokenSource cancellation = new CancellationTokenSource();
+        private readonly ConcurrentDictionary<IModel, ConcurrentDictionary<ulong, TaskCompletionSource<object>>> unconfirmedChannelRequests = new ConcurrentDictionary<IModel, ConcurrentDictionary<ulong, TaskCompletionSource<object>>>();
+
+        public PublishConfirmationListener(IEventBus eventBus)
+        {
+            eventBus.Subscribe<MessageConfirmationEvent>(OnMessageConfirmation);
+            eventBus.Subscribe<PublishChannelCreatedEvent>(OnPublishChannelCreated);
+        }
+
+        public IPublishConfirmationWaiter GetWaiter(IModel model)
+        {
+            var deliveryTag = model.NextPublishSeqNo;
+            var requests = unconfirmedChannelRequests.GetOrAdd(model, new ConcurrentDictionary<ulong, TaskCompletionSource<object>>());
+            var confirmation = new TaskCompletionSource<object>();
+            requests.Add(deliveryTag, confirmation);
+            return new PublishConfirmationWaiter(deliveryTag, confirmation.Task, cancellation.Token, () => requests.Remove(deliveryTag));
+        }
+
+        public void Dispose()
+        {
+            cancellation.Cancel();
+        }
+
+        private void OnMessageConfirmation(MessageConfirmationEvent @event)
+        {
+            ConcurrentDictionary<ulong, TaskCompletionSource<object>> requests;
+            if (!unconfirmedChannelRequests.TryGetValue(@event.Channel, out requests))
+            {
+                return;
+            }
+
+            var deliveryTag = @event.DeliveryTag;
+            var multiple = @event.Multiple;
+            var isNack = @event.IsNack;
+            if (multiple)
+            {
+                // Fix me: ConcurrentDictionary.Keys acquires all locks, it is very expensive operation and could perform slowly.
+                foreach (var sequenceNumber in requests.Keys.Where(x => x <= deliveryTag))
```

Comment:
Actually, it was very hard to dicide what to do in the following situation:
1) IModel BasicAck event fired with delivery tag(let it be x)
2) PublishChannelCreatedEvent was fired because of reconnection, unconfirmedChannelRequests was cleaned, and somebody published message with delivery tag x.

It this scenario, when this two things will happen concurrently, published message could be acked with first fired event, but actually this message is not confirmed by broker.